src/coreclr/tools/Common/Compiler/ObjectWriter/WasmObjectWriter.cs

```diff
@@ -0,0 +1,246 @@
+            {
+                if (!signatureMap.ContainsKey(methodSignatures[i]))
+                {
+                    signatureMap[methodSignatures[i]] = signatureIndex++;
+                }
+            }
+
+            WasmFuncType[] uniqueSignatures = new WasmFuncType[signatureMap.Count];
+            foreach (var kvp in signatureMap)
+            {
+                uniqueSignatures[kvp.Value] = kvp.Key;
+            }
+
+            EmitWasmHeader(outputFileStream);
+            EmitSection(() => WriteTypeSection(uniqueSignatures, logger), outputFileStream, logger);
+            EmitSection(() => WriteFunctionSection(methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
+            EmitSection(() => WriteExportSection(methodBodies.ToArray(), methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
+            EmitSection(() => WriteCodeSection(methodBodies.ToArray(), logger),  outputFileStream, logger);
+        }
+
+        private WasmSection WriteExportSection(IReadOnlyCollection<IMethodBodyNode> methodNodes, IReadOnlyCollection<WasmFuncType> methodSignatures,
+            Dictionary<WasmFuncType, int> signatureMap, Logger logger)
+        {
+            WasmSection exportSection = new WasmSection(WasmSectionType.Export, new SectionData(), "export");
+            SectionWriter writer = exportSection.Writer;
+            // Write the number of exports
+            writer.WriteULEB128((ulong)methodNodes.Count);
+            for (int i = 0; i < methodNodes.Count; i++)
+            {
+                var methodNode = methodNodes.ElementAt(i);
+                var methodSignature = methodSignatures.ElementAt(i);
+                string exportName = methodNode.GetMangledName(_nodeFactory.NameMangler);
+
+                var length = Encoding.UTF8.GetByteCount(exportName);
+                writer.WriteULEB128((ulong)length);
+                writer.WriteUtf8StringNoNull(exportName);
+                writer.WriteByte(0x00); // export kind: function
+                writer.WriteULEB128((ulong)i);
+                if (logger.IsVerbose)
+                {
+                    logger.LogMessage($"Emitting export: {exportName} for function index {i}");
+                }
+            }
+
+            return exportSection;
+        }
+
+        private WasmSection WriteCodeSection(IReadOnlyCollection<IMethodBodyNode> methodBodies, Logger logger)
+        {
+            WasmSection codeSection = new WasmSection(WasmSectionType.Code, new SectionData(), "code");
+
+            SectionWriter writer = codeSection.Writer;
+            // Write the number of functions
+            writer.WriteULEB128((ulong)methodBodies.Count);
+            foreach (IMethodBodyNode methodBody in methodBodies)
+            {
+                ObjectNode objectNode = methodBody as ObjectNode;
+                ObjectNode.ObjectData body = objectNode.GetData(_nodeFactory);
```

Comment:
`as T` followed by `.GetData` without a null check is probably best written as `(ObjectNode)methodBody` instead so we get a cast failure instead of a null reference